plugin/pkg/admission/eventratelimit/admission.go

```diff
@@ -0,0 +1,270 @@
+	eventratelimitapi "k8s.io/kubernetes/plugin/pkg/admission/eventratelimit/apis/eventratelimit"
+	"k8s.io/kubernetes/plugin/pkg/admission/eventratelimit/apis/eventratelimit/validation"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register("EventRateLimit",
+		func(config io.Reader) (admission.Interface, error) {
+			// load the configuration provided (if any)
+			configuration, err := LoadConfiguration(config)
+			if err != nil {
+				return nil, err
+			}
+			// validate the configuration (if any)
+			if configuration != nil {
+				if errs := validation.ValidateConfiguration(configuration); len(errs) != 0 {
+					return nil, errs.ToAggregate()
+				}
+			}
+			return newEventRateLimit(configuration)
+		})
+}
+
+// eventRateLimitAdmission implements an admission controller that can enforce event rate limits
+type eventRateLimitAdmission struct {
+	sync.RWMutex
+	*admission.Handler
+	limitEnforcers []*limitEnforcer
+}
+
+// limitEnforcer enforces a single type of event rate limit, such as server, namespace, or source+object
+type limitEnforcer struct {
+	// type of this limit
+	limitType string
+	// factory for creating a rate limiter
+	newRateLimiter func() flowcontrol.RateLimiter
+	// cache for holding the rate limiters
+	cache cache
+	// function for selecting the cache key with which an event is associated
+	keySelector func(admission.Attributes) interface{}
+}
+
+func newLimitEnforcer(config eventratelimitapi.Limit, rlf rateLimiterFactory) (*limitEnforcer, error) {
+	limitType := config.Type
+	newRateLimiter := func() flowcontrol.RateLimiter {
+		return rlf(config.QPS, config.Burst)
+	}
+	limitEnforcer := &limitEnforcer{
+		limitType:      limitType,
+		newRateLimiter: newRateLimiter,
+	}
+	switch t := config.Type; t {
+	case eventratelimitapi.ServerLimitType:
+		limitEnforcer.cache = newSingleCache(newRateLimiter())
+		limitEnforcer.keySelector = getServerKey
+	case eventratelimitapi.NamespaceLimitType:
+		limitEnforcer.cache = newLRUCache(config.CacheSize)
+		limitEnforcer.keySelector = getNamespaceKey
+	case eventratelimitapi.UserLimitType:
+		limitEnforcer.cache = newLRUCache(config.CacheSize)
+		limitEnforcer.keySelector = getUserKey
+	case eventratelimitapi.SourceObjectLimitType:
+		limitEnforcer.cache = newLRUCache(config.CacheSize)
+		limitEnforcer.keySelector = getSourceObjectKey
+	default:
+		return nil, errors.New(fmt.Sprintf("unknown event rate limit type: %v", t))
+	}
+	return limitEnforcer, nil
+}
+
+func (enforcer *limitEnforcer) accept(attr admission.Attributes) error {
+	key := enforcer.keySelector(attr)
+
+	// do we have a record of similar events in our cache?
+	rateLimiter, found := enforcer.cache.get(key)
+
+	// verify we have a rate limiter for this record
+	if !found {
+		rateLimiter = enforcer.newRateLimiter()
+	}
+
+	// ensure we have available rate
+	filter := rateLimiter.TryAccept()
+
+	// update the cache
+	enforcer.cache.add(key, rateLimiter)
```

Comment:
This looks odd.  We have to do a write every `accept` call?  Did this belong in that `!found` block above?

Reply:
It seemed odd to me, too. It was copied over from the event rate limiting code in the client-side event recorder. I had assumed that Add was needed to make the key the most recently used. I see from the lru code, though, that Get does that already, which makes sense.